test_lift_command flakes on ubuntu CI: Ghidra exits 0 having written nothing

## What happened

`test_lift_command` failed on `ubuntu-latest` during [PR #53](https://github.com/tamada/oinkie/pull/53):

```
thread 'test_lift_command' panicked at tests/cli_test.rs:32:12:
code=2
stderr="Error: Parse error: Expected Ghidra to generate
        \"/tmp/oinkie_workx1qvkv/hello_clang.json\", but it was not found."
```

Re-running the same job with **no code change** passed. `macOS-latest` passed both times.

## It is not the PR

The refactor in #53 was suspected first, so the command line it builds was compared against the one the parent commit builds, for the same input, with the temporary-directory randomness normalised. They are byte-identical:

```
cd "<tmp>" && "<ghidra>/support/analyzeHeadless" "<tmp>" "hello_clang" \
  -import "<abs input>" -scriptPath "<tmp>" -postScript "HighPCodeLifter.java"
```

## The likely cause

`tests/cli_test.rs` has three tests that each spawn a real Ghidra headless run, and cargo runs them in parallel — three JVMs on a two-core runner. In the failing run `test_lift_command` took 17.7s while the other two finished in ~5s; in the passing runs all three take ~24s. The interleaving differs run to run.

`analyzeHeadless` **exits 0 even when its post-script throws**, so a JVM that ran out of memory or was starved reports success and writes nothing, which is exactly the observed shape.

## Partly addressed

#53 now includes the tool's own stdout/stderr (last 20 lines) in the missing-output error, so the next occurrence should say why rather than only naming an absent path in a temporary directory that no longer exists. That makes it diagnosable; it does not stop it happening.

## Options

- Run the lift tests serially — a shared lock, or `--test-threads=1` for `cli_test`. Costs wall time on a test that already dominates the suite.
- Mark them `#[ignore]` and run them in a separate CI step. Keeps the fast suite fast, but they stop being run by `cargo test`.
- Lift once in a fixture shared by the three tests. They differ only in the `-i` handling, so two of them could assert against a shared run.

Related to #24, which is about `GHIDRA_HOME` races in the unit tests, but a different mechanism: this one is resource contention between real Ghidra processes.

## 実際の原因が判明しました（推測は外れです）

#53 に「成果物が無いときはツール自身の出力を持つ」変更を入れたところ、次の失敗が理由を吐きました。並列 JVM のメモリ枯渇ではありません。

```
Caused by: java.util.zip.ZipException: invalid block type
    at java.util.zip.InflaterInputStream.read
    at ghidra.program.model.pcode.LinkedByteBuffer.readPage
    at ghidra.program.model.pcode.PackedDecode.ingestStream
    at ghidra.pcode.utils.SlaFormat.buildDecoder
    at ghidra.app.plugin.processors.sleigh.SleighLanguage.initialize
    at ghidra.app.plugin.processors.sleigh.SleighLanguageProvider.getLanguage
ERROR REPORT: Import failed for file: .../testdata/hello_world/bin/hello_clang
```

Ghidra は SLEIGH の言語定義を `.slaspec` から `.sla` へ**初回使用時にコンパイルし、インストールディレクトリに書き込んでキャッシュ**します。CI では毎回 Ghidra を新規インストールするので、`.sla` は常に未生成の状態から始まります。

そこへ `analyzeHeadless` が 3 プロセス同時に立ち上がり、**同じ `.sla` ファイルの生成を競合**します。負けた側が書きかけのファイルを読み、伸長に失敗する — これが `ZipException: invalid block type` です。共有ファイルの競合であって、リソース不足ではありませんでした。

## テストだけの問題ではありません

`perform_lift` は `opts.iter().par_bridge()` で lift を並列実行します。つまり **Ghidra をインストールした直後のユーザが複数バイナリを一度に lift すると、同じ競合に当たり得ます**。2 回目以降はキャッシュが温まっているので起きません。影響の順序は「CI は毎回危険、ユーザはインストール直後の初回のみ危険」です。

## macOS が緑だったのは無関係でした

ワークフローを読み直したところ、**テストを実行しているのは ubuntu だけ**です（`Convert coverage format to lcov` ステップが `if: matrix.os == 'ubuntu-latest'`）。macOS は fmt/clippy/build しかしていません。「Linux 固有」ではなく「テストを走らせている唯一の場所」でした。

## 対策の方向

- **CI**: テスト前に lift を 1 回直列で走らせてキャッシュを温める。根本の仕組みに効き、並列性も保てます。
- **oinkie 側**: 初回 lift を直列化する、あるいはファイルロックを取る。こちらは設計判断が要るので別途。

--- a/.github/release-notes/v0.4.0.md
+++ b/.github/release-notes/v0.4.0.md
@@ -232,6 +232,9 @@ now is.
 - **#45** — the lifted file decides which operation type reads it, so the
   pipeline no longer assumes P-Code; a representation with no reader is refused
   by name
+- **#54** — the tests that run a real Ghidra no longer run at the same time as
+  each other, so they cannot corrupt its language cache; two of them that
+  differed only in their `-i` path are now one
 - **#55** — `lift` is serial by default, with `-j/--jobs N` to opt back into
   parallelism; the old default could corrupt Ghidra's language cache on a fresh
   installation, and the corruption was reported as a missing output

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -38,21 +38,6 @@ jobs:
           cargo clippy -- -D warnings
           cargo build --release
 
-      # Ghidra compiles its SLEIGH language definitions from .slaspec to .sla
-      # on first use and caches them inside its own installation, which the
-      # runner unpacks fresh every time. The lift tests start three
-      # analyzeHeadless processes at once, and against a cold installation
-      # they race to write the same .sla; the loser reads a half-written file
-      # and dies with "ZipException: invalid block type" -- while
-      # analyzeHeadless still exits 0, so the failure surfaced only as a
-      # missing output file. One serial lift builds the cache before anything
-      # runs in parallel.
-      - name: Warm Ghidra's language cache
-        if: matrix.os == 'ubuntu-latest'
-        run: |
-          export GHIDRA_HOME=${{ env.GHIDRA_INSTALL_DIR }}
-          ./target/release/oinkie lift -d "$RUNNER_TEMP/warm" testdata/hello_world/bin/hello_clang
-
       - name: Convert coverage format to lcov
         if: matrix.os == 'ubuntu-latest'
         run: |

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -35,6 +35,7 @@ path = "cli/main.rs"
 [dev-dependencies]
 assert_cmd = "2.2.2"
 predicates = "3.1.4"
+serial_test = "4.0.1"
 
 [[bin]]
 name = "gencomp"

--- a/tests/cli_test.rs
+++ b/tests/cli_test.rs
@@ -1,7 +1,19 @@
 use assert_cmd::Command;
 use predicates::prelude::*;
+use serial_test::serial;
 use std::fs;
 use tempfile::tempdir;
+
+// Tests marked `#[serial(ghidra)]` start a real Ghidra. They are serialised
+// against each other because Ghidra compiles its SLEIGH language definitions
+// on first use and caches them *inside its own installation*, so two headless
+// runs against an installation nobody has used yet write the same file at the
+// same time and the loser reads a half-written one. analyzeHeadless exits
+// successfully regardless, so it arrives as a missing output rather than an
+// error (#54).
+//
+// The group is named for what is shared rather than left anonymous, so that a
+// future test needing the same exclusion knows which one to join.
 
 #[test]
 fn test_info_command() {
@@ -15,6 +27,7 @@ fn test_info_command() {
 }
 
 #[test]
+#[serial(ghidra)]
 fn test_lift_command() {
     let temp_dir = tempdir().unwrap();
     let dest = temp_dir.path().join("lifted");
@@ -196,8 +209,26 @@ fn test_reaggregate_command() {
 /// also used as the working directory of the Ghidra process. A relative path
 /// must therefore be resolved before the process starts, or Ghidra resolves it
 /// a second time against itself and looks for `irs/irs`.
+/// Both `-i` regressions in one run, because each run is a whole decompiler.
+///
+/// A path given to `-i` is created rather than required to exist, as every
+/// other destination directory in the CLI is and as the temporary directory
+/// used without `-i` is by construction. And it is resolved once: it used to
+/// be resolved twice, by us and again by Ghidra against its own working
+/// directory, so `-i irs` went looking for `irs/irs`.
+///
+/// A relative path that does not exist yet covers both at once, and being
+/// nested covers creating intermediate levels rather than just the last.
+///
+/// Either regression makes the run itself fail here, which is how #36 was
+/// reported -- Ghidra complaining that `irs/irs` did not exist -- rather than
+/// showing up as a stray directory. Both were checked by reverting each fix in
+/// turn. The explicit assertions below still earn their place: they name which
+/// of the two broke, and they catch a variant that doubles the path without
+/// failing outright.
 #[test]
-fn test_lift_command_with_relative_intermediate_dir() {
+#[serial(ghidra)]
+fn test_lift_command_intermediate_dir_is_created_and_resolved_once() {
     // Ghidra rejects any path element starting with '.', and tempdir() names
     // its directories ".tmpXXXX", so the project location needs a plain prefix.
     let temp_dir = tempfile::Builder::new()
@@ -206,52 +237,24 @@ fn test_lift_command_with_relative_intermediate_dir() {
         .unwrap();
     let input = fs::canonicalize("testdata/hello_world/bin/hello_clang").unwrap();
     let dest = temp_dir.path().join("lifted");
-    fs::create_dir(temp_dir.path().join("irs")).unwrap();
 
     Command::cargo_bin("oinkie")
         .unwrap()
         .current_dir(temp_dir.path())
         .arg("lift")
         .arg("-i")
-        .arg("irs")
+        .arg("irs/nested")
         .arg("-d")
         .arg(&dest)
         .arg(&input)
         .assert()
         .success();
 
+    let intermediate = temp_dir.path().join("irs/nested");
+    assert!(intermediate.is_dir(), "the -i directory was not created");
     assert!(
-        !temp_dir.path().join("irs").join("irs").exists(),
+        !intermediate.join("irs/nested").exists(),
         "the -i path was resolved twice"
     );
-    assert!(dest.join("hello_clang.json").exists());
-}
-
-/// Every other destination directory in the CLI is created for the user, and
-/// the temporary directory used when `-i` is omitted exists by construction.
-/// A path passed to `-i` should be no different.
-#[test]
-fn test_lift_command_creates_the_intermediate_dir() {
-    // See the note above: the project location must not sit under a dot-directory.
-    let temp_dir = tempfile::Builder::new()
-        .prefix("oinkie_test")
-        .tempdir()
-        .unwrap();
-    let input = fs::canonicalize("testdata/hello_world/bin/hello_clang").unwrap();
-    let dest = temp_dir.path().join("lifted");
-    let intermediate = temp_dir.path().join("does/not/exist/yet");
-
-    Command::cargo_bin("oinkie")
-        .unwrap()
-        .arg("lift")
-        .arg("-i")
-        .arg(&intermediate)
-        .arg("-d")
-        .arg(&dest)
-        .arg(&input)
-        .assert()
-        .success();
-
-    assert!(intermediate.is_dir(), "the -i directory was not created");
     assert!(dest.join("hello_clang.json").exists());
 }

